### PR TITLE
[incubator-kie-issues-1517] Add Transactional Rest endpoints to UserTasks

### PR DESCRIPTION
**Pull request by elguardian**
Closes: https://github.com/apache/incubator-kie-issues/issues/1517

### CODE REVIEW

**Comment by kie-ci3** [comment]
**PR job** `#1` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-kogito-runtimes -u https://github.com/apache/incubator-kie-kogito-runtimes/pull/3701 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3701/1/display/redirect

**Test results:**
- PASSED: 3375
- FAILED: 3

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3701/1/testReport/org.jbpm.bpmn2/IntermediateEventTest/testTimerBoundaryEventInterrupting/">org.jbpm.bpmn2.IntermediateEventTest.testTimerBoundaryEventInterrupting</a></summary>
<br/>expected: 2<br/> but was: 1
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3701/1/testReport/org.jbpm.bpmn2/IntermediateEventTest/testIntermediateCatchEventTimerDuration/">org.jbpm.bpmn2.IntermediateEventTest.testIntermediateCatchEventTimerDuration</a></summary>
<br/>expected: 2<br/> but was: 1
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3701/1/testReport/org.jbpm.bpmn2/IntermediateEventTest/testEventBasedSplit2/">org.jbpm.bpmn2.IntermediateEventTest.testEventBasedSplit2</a></summary>
<br/>expected: 2<br/> but was: 1
</details>

File: kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/process/util/CodegenUtil.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        return String.format("kogito.%s.%s", generator.name(), propertyName);
+    }
+
+    public static String globalProperty(String propertyName) {
+        return String.format("kogito.%s", propertyName);
+    }
+
+    /**
+     * There is a preference order about how to compute transaction enabling.
+     * 1. We check the property for the custom generator and
```
Comment (martinweiler) [inline kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/process/util/CodegenUtil.java:46]: Comment should be finished to describe the logic which properties are taken into account.

File: kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/process/util/CodegenUtil.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+     */
+    public static boolean isTransactionEnabled(Generator generator, KogitoBuildContext context) {
+        return getProperty(generator, context, TRANSACTION_ENABLED, Boolean::parseBoolean, true);
+    }
+
+    /**
+     * There is a preference order about how to compute transaction enabling.
+     * 1. We check the property for the custom generator and
```
Comment (martinweiler) [inline kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/process/util/CodegenUtil.java:54]: Same as above (or just add the comment in one place)

File: kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/usertask/UserTaskCodegen.java [hunk 3]
```diff
@@ -99,10 +101,11 @@ public class UserTaskCodegen extends AbstractGenerator {
 
     public static final String SECTION_CLASS_NAME = "usertask";
 
-    TemplatedGenerator templateGenerator;
+    private TemplatedGenerator templateGenerator;
     private List<Work> descriptors;
     private TemplatedGenerator producerTemplateGenerator;
     private TemplatedGenerator restTemplateGenerator;
+    boolean transactionEnabled;
```
Comment (martinweiler) [inline kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/usertask/UserTaskCodegen.java:108]: This field does not seem to be used anywhere and can be removed.

**Review by martinweiler — COMMENTED** [review]
Question regarding `UserTaskCodegen`. Running an example with `kogito.processes.transactionEnabled=true`, I can see `@Transactional` annotations in the following generated file:
```
@ApplicationScoped
public class UserTasksServiceProducer {

    @Produces
    @jakarta.transaction.Transactional()
    public UserTaskService userTaskService(Application application) {
        return new UserTaskServiceImpl(application);
    }
```
What I was expecting to see instead was `@Transactional` annotation added to `UserTasksResource`:
```
@Path("/usertasks/instance")
public class UserTasksResource {

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    public List<UserTaskView> list(@QueryParam("user") String user, @QueryParam("group") List<String> groups) {
        return userTaskService.list(IdentityProviders.of(user, groups));
    }
```
Am I missing something?

**Comment by elguardian** [comment]
Hi Martin, in theory should be the custom property kogito.usertasks.transactionEnabled=true or kogito.transactionEnabled=true or nothing at all

File: kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/process/util/CodegenUtil.java (hunk 1 above)
Comment (elguardian) [reply kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/process/util/CodegenUtil.java]: fixed

File: kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/process/util/CodegenUtil.java (hunk 2 above)
Comment (elguardian) [reply kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/process/util/CodegenUtil.java]: fixed

File: kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/usertask/UserTaskCodegen.java (hunk 3 above)
Comment (elguardian) [reply kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/usertask/UserTaskCodegen.java]: fixed

**Comment by kie-ci3** [comment]
**PR job** `#2` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-kogito-runtimes -u https://github.com/apache/incubator-kie-kogito-runtimes/pull/3701 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3701/2/display/redirect

**Test results:**
- PASSED: 3379
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3701/2/testReport/org.jbpm.bpmn2/IntermediateEventTest/testEventBasedSplit2/">org.jbpm.bpmn2.IntermediateEventTest.testEventBasedSplit2</a></summary>
<br/>expected: 2<br/> but was: 1
</details>

File: kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/process/util/CodegenUtil.java [hunk 4]
```diff
@@ -61,7 +65,9 @@ public static String globalProperty(String propertyName) {
      * @see CodegenUtil#getProperty
      */
     public static boolean isTransactionEnabled(Generator generator, KogitoBuildContext context) {
-        return getProperty(generator, context, TRANSACTION_ENABLED, Boolean::parseBoolean, true);
+        boolean propertyValue = getProperty(generator, context, TRANSACTION_ENABLED, Boolean::parseBoolean, true);
+        LOG.info("trying to compute property {} for generator {} property with value {}", generator.name(), TRANSACTION_ENABLED, propertyValue);
```
Comment (martinweiler) [inline kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/process/util/CodegenUtil.java:69]: Just a nit-pick, `generator.name()` and `TRANSACTION_ENABLED` should be switched around

**Review by martinweiler — APPROVED** [review]
LGTM, thank you @elguardian

File: kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/process/util/CodegenUtil.java (hunk 4 above)
Comment (elguardian) [reply kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/process/util/CodegenUtil.java]: fixed

**Comment by elguardian** [comment]
python failure realted to env

**Comment by kie-ci3** [comment]
**PR job** `#3` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-kogito-runtimes -u https://github.com/apache/incubator-kie-kogito-runtimes/pull/3701 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3701/3/display/redirect

**Test results:**
- PASSED: 3378
- FAILED: 3

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3701/3/testReport/org.jbpm.bpmn2/IntermediateEventTest/testTimerBoundaryEventInterrupting/">org.jbpm.bpmn2.IntermediateEventTest.testTimerBoundaryEventInterrupting</a></summary>
<br/>expected: 2<br/> but was: 1
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3701/3/testReport/org.jbpm.bpmn2/IntermediateEventTest/testIntermediateCatchEventTimerDuration/">org.jbpm.bpmn2.IntermediateEventTest.testIntermediateCatchEventTimerDuration</a></summary>
<br/>expected: 2<br/> but was: 1
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3701/3/testReport/org.jbpm.bpmn2/IntermediateEventTest/testEventBasedSplit2/">org.jbpm.bpmn2.IntermediateEventTest.testEventBasedSplit2</a></summary>
<br/>expected: 2<br/> but was: 1
</details>

**Pull request merged**